Повелитель данных

kind-chigik opened the pull request:
---
:mortar_board: [Повелитель данных](https://up.htmlacademy.ru/yii/1/user/669277/tasks/8)

kind-chigik committed:
Созданы sql файлы для заполнения таблиц БД

kind-chigik committed:
Добавлен сценарий для теста класса DataImporter

kind-chigik committed:
Изменены названия полей

kind-chigik committed:
Скорректирован код проверки исключения, исправлено описание к функции areArraysEqual

kind-chigik committed:
Добавлен датасет для таблицы специализаций

kind-chigik committed:
Скорректированы названия полей

Для корректного импорта данных из csv файлов в sql запросы название полей в файлах должно совпадать с названием полей в таблице. С этой целью были скорректированы названия полей и добавлены новые поля.
Также была пересмотрена обязательность заполнения некоторых полей.

kind-chigik committed:
Добавлен класс для импорта данных в таблицы БД

kind-chigik committed:
Ничего не менялось

Файл случайно был переписан в процессе теста класса импорта данных в таблицы БД. Содержимое было восстановлено и более не отличается от первоначального варианта

kind-chigik committed:
Изменена кодировка на UTF-8

kind-chigik committed:
Устранены ошибки в запросах из-за неверной кодировки

kind-chigik committed:
Удалены лишние знаки

kind-chigik committed:
Изменено условие об обязательности заполнения у одного из полей

Ovsyanka commented on schema.sql:
```diff
@@ -86,7 +86,7 @@ CREATE TABLE replies (
   id INT UNSIGNED AUTO_INCREMENT PRIMARY KEY,
   title VARCHAR(128) NOT NULL UNIQUE,
-  category_id INT UNSIGNED NOT NULL,
+  category_id INT UNSIGNED,
```
Сначала удивился необязательному полю, а потом понял, что не понимаю, что это вообще за таблица. В ТЗ про специализации я вижу следующее: "Под заголовком «Специализации» в виде ссылок отображаются выбранные исполнителем категории.". То есть "специализации" это "категории", которые выбрал пользователь. То есть в таблице специализаций должны хранится соответствия между пользователями и категориями, которые они выбрали. И я бы назвал эту таблицу user_categories вместо specializations, чтобы было понятно по названию, что она содержит отношения между этими двумя сущностями.

Ovsyanka commented on schema.sql:
```diff
@@ -5,31 +5,33 @@ USE taskforce;
   avatar_path VARCHAR(128),
-  birthday DATETIME,
-  description TEXT NOT NULL,
+  bd DATETIME,
```
birthday было лучше гораздо, на мой взгляд. А так у меня, например, первая мысль - bd = База Данных. Потом наступает когнитивный диссонанс потому что по-английски это должно было быть db (DataBase), а не bd.
Не принципиально, решать тебе.

Ovsyanka commented on schema.sql:
```diff
@@ -5,31 +5,33 @@ USE taskforce;
   skype VARCHAR(64),
   another_contact VARCHAR(64),
-  registration_date DATETIME DEFAULT CURRENT_TIMESTAMP,
+  dt_add DATETIME DEFAULT CURRENT_TIMESTAMP,
```
Лучше registration_date (add_date, etc.). Что-то состоящее из слов. dt - само по себе не понятно что это такое. У меня, например, при виде такого сокращения первая мысль - data type какой-то. С date ассоциации не возникает. А у кого-то другого свои могут быть ассоциации.
Не принципиально, решать тебе.

Ovsyanka commented on schema.sql:
```diff
@@ -5,31 +5,33 @@ USE taskforce;
   rank INT DEFAULT '0',
-  city_id INT UNSIGNED NOT NULL,
+  address VARCHAR(128),
+  city_id INT UNSIGNED,
```
В ТЗ я не нашел явного требования, но по-моему предполагается, что человек обязательно какой-то город выбирает, так как не описано как должны себя вести система когда город не выбран (какие показывать объявления - все/никакие). Возможно, я этого не заметил - тогда поправь меня.

Ovsyanka commented on schema.sql:
```diff
@@ -42,15 +44,19 @@ CREATE TABLE photos_work (
-  creation_date DATETIME DEFAULT CURRENT_TIMESTAMP,
-  status ENUM('new', 'cancellation', 'in_work', 'completed', 'failed'),
+  budget INT,
+  dt_add DATETIME DEFAULT CURRENT_TIMESTAMP,
```
Насчет названия поля аналогично тому, что раньше.

Ovsyanka commented on schema.sql:
```diff
@@ -68,18 +74,19 @@ CREATE TABLE attachments (
-  reply_text TEXT NOT NULL,
-  creation_time DATETIME DEFAULT CURRENT_TIMESTAMP,
+  description TEXT NOT NULL,
+  dt_add DATETIME DEFAULT CURRENT_TIMESTAMP,
```
Насчет названия поля аналогично тому, что раньше.

Ovsyanka commented on schema.sql:
```diff
@@ -68,18 +74,19 @@ CREATE TABLE attachments (
+  rate INT,
   user_id INT UNSIGNED,
-  task_id INT UNSIGNED NOT NULL,
+  task_id INT UNSIGNED,
```
Разве reply может не относиться ни к какому заданию?

Ovsyanka commented on test.php:
```diff
@@ -111,26 +111,27 @@
 /**
- * Проверяет наличие значений одного массива в другом массиве
+ * Проверяет наличие значений одного массива в другом массиве и выводит сообщение о результате проверки
  * 
  * @param $array1 array Массив с объектами, который должны содержаться в другом массиве
  * @param $array2 array Массив с объектами, в котором должны содержаться искомые значения
  * 
- * @return string Сообщение о результате проверки
+ * @return void
```
Воо, замечательно =)

Ovsyanka commented on testimporter.php:
```diff
@@ -0,0 +1,26 @@
+    $query = $file->convertDataToSql($data, $titlesFields);
+    print $query . '<br>';
+    $sqlFilePath = __DIR__ .'/sql/'. $fileName . '.sql';
+    $sqlFile = $file->createSqlFile($sqlFilePath, $query);
```
> Реализуйте класс для перевода данных из CSV в SQL формат для одного файла.

В ТЗ не сказано, что он должен предоставлять функционал для получения частей этих csv данных (getData, getTitlesFilds) или преобразовывать данные из массивов в sql (convertDataToSql). Всё это - внутренняя реализация задачи перевода csv в sql. Так что эти методы должны быть приватными и вызываться внутри класса. Работа с ним будет выглядеть примерно так:

```php
$csvToSqlConverter = new CsvToSqlConverter($tableName, $csvFilePath);
$sqlFilePath = $outSqlDir . DIRECTORY_SEPARATOR . "{$tableName}.sql";
$csvToSqlConverter->createSqlFile($sqlFilePath);
```

Ovsyanka commented on src/DataImporter.php:
```diff
@@ -0,0 +1,77 @@
+
+class DataImporter {
+    private $file;
+    private $fileName;
```
Называется fileName, а по сути то это имя таблицы. Надо переименовать в tableName.

Ovsyanka commented on src/DataImporter.php:
```diff
@@ -0,0 +1,77 @@
+    
+    public function getData() : array
+    {
+        $this->file->setFlags(\SplFileObject::READ_CSV | \SplFileObject::SKIP_EMPTY | \SplFileObject::READ_AHEAD);
```
Это лучше смотрится в конструкторе, на мой взгляд. Не принципиально.

Ovsyanka commented on src/DataImporter.php:
```diff
@@ -0,0 +1,77 @@
+        return $result;
+    }
+
+    public function getTitlesFilds() {
```
Обычно, заголовки в таком контексте называются headers. То есть эту функцию корректней назвать getCsvHeaders

Ovsyanka commented on src/DataImporter.php:
```diff
@@ -0,0 +1,77 @@
+        return $titlesFields;
+    }
+
+    public function convertDataToSql(array $data, array $titlesFields) : string
```
На мой взгляд было бы лучше, если бы в $data не было заголовков, раз уж они передаются отдельно. Это довольно неожиданно, что они должны быть переданы и в data и в titleFields.

Ovsyanka commented on src/DataImporter.php:
```diff
@@ -0,0 +1,77 @@
+
+    public function createSqlFile(string $sqlFilePath, string $query) 
+    {
+        $sqlFile = new \SplFileObject($sqlFilePath, 'w+');
```
Все используемые классы стоит подключать, используя use. Чтобы зависимости модуля были сразу видны. Мое личное мнение - исправлять не обязательно.

Ovsyanka commented on test.php:
```diff
@@ -111,26 +111,27 @@
 }
 catch (FakeStatusException $s) {
-    print 'Ошибка: ' . $s->getMessage();
+    print 'Ошибка: ' . $s->getMessage() . '<br>';
```
Если я увижу при выполнении теста запись "Ошибка: ...." я буду в полной уверенности, что тест провален. А тут ситуация противоположная =) Тесты должны выдавать результат проверок в унифицированном виде, чтобы не было вопросов - пройден тест или провален. Для остальных тестов ты пишешь "проверка прошла успешно / провалена" - так же нужно писать и в этом.

Ovsyanka reviewed (commented):
Название сценария для конвертации не отражает его действие. Он же не импортирует данные, а конвертирует. Так что его переименовать надо. _Вообще, было бы неплохо написать также и импортирование данных, чтобы не было нужды делать это сторонними средствами (руками), но об этом в ТЗ ничего нет - так что целиком на твое усмотрение._ 

Еще, test на мой взгляд тут не в тему совсем (да, в ТЗ написано "тестовый" сценарий, но он вовсе и не тестовый - он же не тестирует ничего). Так что я бы от этой приписки избавился, но ты сама смотри.

Ovsyanka reviewed (commented):
В нашем случае sql файлы являются производными данными от csv файлов. Такие данные не следует хранить в репозитории. Учитывая, что это учебный проект, можешь действовать по своему выбору - либо следить за их актуальностью, либо удалить из репозитория и добавить в .gitignore.

kind-chigik committed:
Изменены заголовки столбцов, добавлены новые столбцы и данные

kind-chigik committed:
Изменено название столбца на более понятное

kind-chigik committed:
Скорректированы названия полей таблиц, добавлены новые данные

kind-chigik committed:
Удален, т.к. удалена таблица специализаций

kind-chigik committed:
Переписано сообщение, которое выводится по результату проверки исключений

kind-chigik committed:
Правки названий полей, обязательности заполнения, таблиц

- Названия некоторых полей переписаны на более понятные
- Некоторые поля помечены как обязательные для заполнения
- Удалена таблица специализаций, т.к. была лишней.

kind-chigik committed:
Переименован сценарий и сокращен код работы с классом DataConverter

kind-chigik committed:
Переименован сценарий и скорректирован код класса

- Теперь задачи по импорту данных из csv файлов и составлению sql запросов выполняются внутри класса. Для этого были созданы дополнительные переменные, которые опредяются в конструкторе.

- Видимость методов, которые не используются за пределами класса, изменена на приватную.

- Изменены названия некоторых свойств и методов

kind-chigik committed:
Переименован на сonvertertosql

Ovsyanka commented on schema.sql:
```diff
@@ -75,27 +75,20 @@ CREATE TABLE attachments (
+CREATE TABLE user_categories (
   user_id INT UNSIGNED,
-  specialization_id INT UNSIGNED,
-  PRIMARY KEY (user_id, specialization_id),
+  category_id INT UNSIGNED,
```
not null

Ovsyanka commented on test.php:
```diff
@@ -114,15 +114,15 @@
 }
 catch (FakeStatusException $s) {
-    print 'Ошибка: ' . $s->getMessage() . '<br>';
+    print 'Проверка для исключения: ' . $s->getMessage() . ' - прошла успешно' . '<br>';
```
А если не успешно?

Ovsyanka reviewed (approved):
Замечательно!
Есть пара не критичных замечаний, их вполне можно поправить позже.
